balancer/xds/internal/regenerate_scripts/utils/cleanup.sh

```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+for f in $(find . -name \*.proto)
+do
+  commands=(
+    # Import mangling.
+    -e 's#import "gogoproto/gogo.proto";##'
+    # -e 's#import "validate/validate.proto";##'
+    # Remove references to gogo.proto extensions.
+    -e 's#option (gogoproto\.[a-z_]\+) = \(true\|false\);##'
+    -e 's#\(, \)\?(gogoproto\.[a-z_]\+) = \(true\|false\),\?##'
+    # gogoproto removal can result in empty brackets.
+    -e 's# \[\]##'
+    # gogoproto removal can result in four spaces on a line by itself.
+    -e '/^    $/d'
+  )
+  sed -i "${commands[@]}" "$f"
+
+  # gogoproto removal can leave a comma on the last element in a list.
+  # This needs to run separately after all the commands above have finished
+  # since it is multi-line and rewrites the output of the above patterns.
+  sed -i -e '$!N; s#\(.*\),\([[:space:]]*\];\)#\1\2#; t; P; D;' "$f"
```

Comment:
- I think the "t" is ignored completely because there is no label?  It has no effect AFAICT anyway.
- I don't think the $ and ! do anything.

Another way of doing this:

```
sed -e ':a N; s#\(.*\),\([[:space:]]*\];\)#\1\2#; ba' "$f"
```

This will catch the following pattern (not sure if it's helpful though):

```
a,
b,

]
```

Reply:
OK, I've refamiliarized myself with `sed` a bit more and see what's going on.  The `$!N` is necessary for POSIX `sed` to not exit when `N` is executed on the last line (GNU `sed` does not follow this behavior; it prints the final line and exits).  And the `t` with no label jumps to the end of the script (restarting at the beginning for the next line with an empty pattern buffer) when the regex matches.

So what you have is fine.  Or you can do something a little more complicated to catch `,[[:space:]]*];` patterns spanning more than two adjacent lines:

```
sed -e ':a; /,[[:space:]]*$/ { $!N; s#\(.*\),\([[:space:]]*];\)#\1\2#; T a }' "$f"
```

(Note that escaping a close bracket is unnecessary.)